crates/kilt-dip-primitives/src/merkle/v0/provider_state/tests.rs

```diff
@@ -0,0 +1,566 @@
+				hex!("9f0b3c252fcb29d88eff4f3de5de4476c3ffbf805254dc9131b269f3bbbb71f58a76a5034b2bc2faaab0d1cf45c3819dc6e69740804bc059c3d96f627e09a3b6c0f9851d902f84ac68006617289ac0b7d0a272b36280d97e2394406f94be4266da29b6fe7f3178059525eaf3c9b540064389af020bf180636959b43018d3ff8a55246d5874a16c93e85bd2a58c82ebfc1b54dd9b2a7d0780d3c1a10188200f31459d722f7efc693736d1a36af5644fd949b2e411d7942597800328f24d0485b9701135913a569f6ccbf261a05d055183abf3e4ecb4e4375b7c80f3229cd59de7b1e604f110cbcf814466f2d2973e9bdb6c106a662c576e0820e480b66b29cbd45f93602dbc9f1175407c6f69bd686d23dd22a8f0dfe9cff08843ad80ddb2d426c0c546068b429e77253e0a8a32e818151f5fc031e899a0f6acad157580ea7fb3cad8e128cc295194658016f4865ef37501e5759fb4f15cb2ecb689e85e80e9f3cac1b25842da7fbaf947952dc30329a1d19037ab21baed3851acbee629f6800d898e2a4a6ee9969a233c4741e4441c0fe393104b3cfc5adcf348f3ef20fc7480ac6c622536e593ae3c9d423a461faafc7abbf01ecb129e69d66f3382eaf484dc80c7ba3cadffaea5acd013dba51c96129ae93ea6cd45f3930e4302f5b100f6deae806f29f805e30029363e42381d6609ecb6837411bd6fd676c0a37621a3b5588101").to_vec()
+			].into_iter().into(),
+		};
+		// Only interested in the parachain head verification part, we skip everything
+		// else.
+		let proof = ParachainDipDidProof::<_, (), (), _, (), (), ()>::with_provider_head_proof(provider_head_proof);
+		let proof_verification_result = proof
+			.verify_provider_head_proof_with_state_root::<BlakeTwo256, HeaderFor<SpiritnetRuntime>>(
+				2_086,
+				&relay_state_root,
+			)
+			.unwrap();
+		assert_eq!(
+			proof_verification_result.state_root,
+			hex!("2937179bfe939750c770f2aa6e84c6b8cf9d0aa9ab852243ceb78e3eeb93fc56").into()
+		);
+	}
+
+	#[test]
+	fn verify_provider_head_proof_with_state_root_wrong_relay_hasher() {
+		// Storage proof generated at Polkadot block `19_663_508` with hash
+		// `0x6e87866fb4f412e1e691e25d294019a7695d5a756ee7bc8d012c25177b5e1e13` for
+		// storage key
+		// `0xcd710b30bd2eab0352ddcc26417aa1941b3c252fcb29d88eff4f3de5de4476c32c0cfd6c23b92a7826080000`
+		// (`paras::heads(2_086)`)
+		let relay_state_root: H256 = hex!("623b36bddae282e9fefab4707697171a594fdb27e90fd4ada4ebcc356438b070").into();
+		let provider_head_proof = ProviderHeadStateProof {
+			relay_block_number: 19_663_508,
+			proof: vec![
+				hex!("560cfd6c23b92a7826080000f102e902fb1bd938b2f4fcea70641da8e64e0e11098b92b767279227cdfdc0ae9500da99d27e5f012937179bfe939750c770f2aa6e84c6b8cf9d0aa9ab852243ceb78e3eeb93fc56eacc28c5503a155c4d8bc7ee4797c38e212428cefff0a7ad19b28ebbab793e64080661757261204a207d0800000000056175726101017cda19117c87384aaebfd2ae546771bcfbfe7011a91119932883382cc62be3050d745c9734f422228c7c43d87e6172519019829b14b2d3b64afafb1fb7d3a683").to_vec(),
```

Comment:
Yes, exactly. The vector is just the ordered list of the tree leaves. They are then recursively hashed until we get to the root.

Reply:
Okay I do not exactly get why the `ProviderHeadStateProof` is using the `BoundedBlindedValue`. Please correct me if I am wrong, but the struct doesn't really have a usage or? Would it be not better to use the `StorageProof` struct?

Reply:
It is just a wrapper type to implement the `GetWorstCase` on, without having to rely on `BoundedVec` and const sizes everywhere. But as I said, this is gone in the benchmarking PR, and everything reverted back to simple `Vec`, since the benchmarking logic has changed.